server/service/src/processors/requisition_auto_finalise/requisition_auto_finalise.rs

```diff
@@ -0,0 +1,178 @@
+        "Automatically finalises requisitions according to store preference".to_string()
+    }
+
+    async fn try_process_record(
+        &self,
+        ctx: &ServiceContext,
+        _service_provider: &ServiceProvider,
+        changelog: &ChangelogRow,
+    ) -> Result<Option<String>, ProcessorError> {
+        let connection = &ctx.connection;
+        let store_id = changelog
+            .store_id
+            .clone()
+            .ok_or(ProcessorError::OtherError(
+                "Changelog filter should have ensured a store_id".to_string(),
+            ))?;
+
+        let should_auto_finalise = RequisitionAutoFinalise {}
+            .load(connection, Some(store_id))
+            .map_err(|e| ProcessorError::OtherError(e.to_string()))?;
+
+        if !should_auto_finalise {
+            return Ok(None);
+        }
+
+        let invoice_row_repo = InvoiceRowRepository::new(connection);
+        let invoice = invoice_row_repo
+            .find_one_by_id(&changelog.record_id)?
+            .ok_or(ProcessorError::RecordNotFound(
+                "Invoice".to_string(),
+                changelog.record_id.clone(),
+            ))?;
+
+        if invoice.r#type != InvoiceType::OutboundShipment {
+            return Ok(None);
+        }
+
+        let Some(requisition_id) = invoice.requisition_id else {
+            return Ok(None);
+        };
+
+        if invoice.status != InvoiceStatus::Shipped && invoice.status != InvoiceStatus::Verified {
+            return Ok(None);
+        }
+
+        let requisition_row_repo = RequisitionRowRepository::new(connection);
+        let mut requisition = requisition_row_repo
+            .find_one_by_id(&requisition_id)?
+            .ok_or(ProcessorError::RecordNotFound(
+                "Requisition".to_string(),
+                changelog.record_id.clone(),
+            ))?;
+
+        if requisition.r#type != RequisitionType::Response {
+            return Ok(None);
+        }
+
+        if requisition.status == RequisitionStatus::Finalised {
+            return Ok(None);
+        }
+
+        let invoices = InvoiceRepository::new(connection).query_by_filter(
+            InvoiceFilter::new().requisition_id(EqualFilter::equal_to(&requisition.id)),
+        )?;
+
+        let requisition_lines = RequisitionLineRepository::new(connection).query_by_filter(
+            RequisitionLineFilter::new().requisition_id(EqualFilter::equal_to(&requisition.id)),
+        )?;
+        if requisition_lines.len() == 0 {
+            return Ok(None);
+        }
+
+        let invoice_lines = InvoiceLineRepository::new(connection).query_by_filter(
+            InvoiceLineFilter::new()
+                .invoice_id(EqualFilter::equal_any(
+                    invoices.into_iter().map(|i| i.invoice_row.id).collect(),
+                ))
+                .r#type(EqualFilter {
+                    equal_to: Some(InvoiceLineType::StockOut),
+                    ..Default::default()
+                }),
```

Comment:
Whenever I try to do `EqualFilter::equal_to` the compiler throws a wobbly that the inner type needs to be `String` but I'm giving it `InvoiceLineType`. The generics are getting lost? Anyway rather than fight it I just instantiated the struct instead, which does the generic fine 🤷 .

Reply:
I think going full generic sounds great! I've been kinda confused what's going on in that file for a while. I suspect if we needed specific handling for a particular type the compiler would be smart enough to use that if implemented rather than generic (I presume the compiler generates functions for every type used with the generic anyway, we'd just be manually implementing instead)